### DISCUSSION TITLE
Raise NotImplementedError when I use cart_comm.Neighbor_alltoall(send_buf, recv_buf)

### DISCUSSION THREAD

**Discussion by Wuyuchen2190** [mpi4py/mpi4py#145] [Q&A] [answered]
Hello professor,I have met a problem when I used the method ——cart_comm.Neighbor_alltoall(send_buf, recv_buf) .It raised NotImplementedError.

how should I solve this problem?  Dose it mean the mpi4py package I download from pypi is damaged? 

 I have learned your code followed:
`@unittest.skipIf(not have_feature(), 'mpi-neighbor')  
class BaseTestCCONghBuf(object):
    COMM = MPI.COMM_NULL
    def testNeighborAllgather(self):
        for comm in create_topo_comms(self.COMM):
            rsize, ssize = get_neighbors_count(comm)
            for array, typecode in arrayimpl.subTest(self):
                if unittest.is_mpi_gpu('openmpi', array):
                    # segfault as of OpenMPI 4.1.1; TODO(leofang): why?
                    if array.backend == 'numba':
                        continue
                for v in range(3):
                    sbuf = array( v, typecode, 3)
                    rbuf = array(-1, typecode, (rsize, 3))
                    comm.Neighbor_allgather(sbuf.as_mpi(), rbuf.as_mpi())
                    for value in rbuf.flat:
                        print(value)
            comm.Free()`

But it seems that  'unittest' and 'arrayimpl '  aren't  built-in modules in  mpi4py and I haven't found them in your codes.  

I wanna ask what's the role of "@unittest.skipIf(not have_feature(), 'mpi-neighbor')"? And could you please teach me how to use  cart_comm.Neighbor_allgather(send_buf, recv_buf)  successfully?

my test code is followed:

`import numpy as np
from mpi4py import MPI
comm = MPI.COMM_WORLD
rank = comm.Get_rank()
dims = [3, 3]
periods = [True, True]
cart_comm = comm.Create_cart(dims, periods)
send_buf = np.arange(4, dtype='i')
recv_buf = np.empty(4, dtype='i')
cart_comm.Neighbor_alltoall(send_buf, recv_buf)
print('Neighbor_allgather (periodic): rank {rank} has {recv_buf}')`

and the Traceback message is followed:
`    cart_comm.Neighbor_alltoall(send_buf, recv_buf)#sbuf.as_mpi(), rbuf.as_mpi()

  File "mpi4py\MPI\Comm.pyx", line 2154, in mpi4py.MPI.Topocomm.Neighbor_alltoal
l
NotImplementedError`

Thank you for your time!

**Comment by dalcinl** [comment, marked as answer]
If you get a `NotImplementedError` exception when calling any MPI function or method, that means that you are using a backend MPI implementation that does not support the feature you are trying to use. There is nothing mpi4py can do about it, your MPI does not support the full MPI-3.1 standard (most likely because your MPI is quite old and implements MPI-2.x).

You did not mention it, so I have to ask: what MPI implementation and version are you using? What's the output of `print(MPI.get_vendor())` and `print(MPI.Get_library_version())`? What's your operating system?

**Reply by Wuyuchen2190** [reply]
Thank you professor dalcinl, I am using "Microsoft MPI 10.0.12498.5" , the output is   "('Microsoft MPI', (10, 1, 0)) " and  "Microsoft MPI 10.0.12498.5" respectively. And my operating system is  "Microsoft Windows 10". 

Oh! I got it, I should  change a  MPI implementation,yeah?  Professor, do you have some recomandation for me ?  What's the version should I choose if I use OpenMPI ? 

Ps: When I installed OpenMPI, it raised "configure: WARNING: neither statfs() and statvfs() were found, configure: error: Cannot continue". Professor,do you have any method to solve this problem?  Dose it related to  libibnetdisc.dev  that I have no idea how to get ?

Really really thank you professor!

**Reply by dalcinl** [reply]
As you are running Windows, I don't think you have many options. You could use Intel MPI, I know it supports Windows. But to install mpi4py, you will need a C compiler (MSVC, or perhaps Intel compiler?) to build the package. I have very little Windows expertise, I'm afraid I'll not be able to help you.

@leofang Maybe you have some tips?

**Reply by Wuyuchen2190** [reply]
Thank you so much !

**Reply by leofang** [reply]
No I don't have better advices for Windows users unfortunately.
